File: sdk/reproit-ios/Sources/ReproIt/Exchange.swift
// Production outbound-exchange capture for the iOS SDK.
//
// Port of sdk/reproit-backend-node/instrument.js (and its Rust twin) to
// Foundation. A recorded exchange is what deterministic local replay stubs,
// so responses are captured verbatim up to a fixed inline budget; an
// over-budget body keeps its byte count and sha256 and is marked truncated,
// which makes replay fail closed with a named reason instead of guessing.
//
// The shape is byte-compatible with the backend SDKs:
//
//   {"protocol":"http",
//    "request":{"method","url","headers","body"},
//    "response":{"status","headers","body"}}
//
// This is deliberately NOT the runner marker shape in Causal.swift. That
// marker is a separate, older contract consumed by the fuzz harness and is
// left byte-unchanged; this type is the production capsule contract.

import Foundation

#if canImport(CryptoKit)
  import CryptoKit
#endif

/// Inline body budget per exchange side. Beyond it the body is dropped and
/// only provable identity (byte count + sha256) remains. Matches
/// `MAX_EXCHANGE_BODY_BYTES` in instrument.js.
let reproitMaxExchangeBodyBytes = 8 * 1024

/// Recorded headers per side, capped to keep events bounded. Matches
/// `MAX_EXCHANGE_HEADERS` in instrument.js.
let reproitMaxExchangeHeaders = 32

/// Secret-shaped field names, redacted at source before anything leaves the
/// device. Byte-identical to `SECRET_PARTS` in the Node SDK's index.js.
private let reproitSecretParts = [
  "password", "passwd", "secret", "token", "authorization", "cookie", "email", "phone",
  "apikey", "publishablekey", "privatekey", "accesskey", "signingkey", "idempotencykey",
]

/// True when a field name looks secret. Non-alphanumerics are stripped and the
/// name lowercased first, so `private-key`, `Access Key`, and `signingKey` all
/// match, exactly like `secretField` in the Node SDK.
func reproitIsSecretField(_ name: String) -> Bool {
  let folded = name.lowercased().filter { $0.isLetter || $0.isNumber }
  return reproitSecretParts.contains { folded.contains($0) }
}

/// Full lowercase-hex sha256 of `data`, or nil when no crypto is available.
/// A nil result never becomes a silent omission: the caller marks the body
/// truncated regardless, and replay fails closed on truncated bodies.
func reproitSha256Hex(_ data: Data) -> String? {
  #if canImport(CryptoKit)
    return SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
  #else
    return nil
  #endif
}

/// The `$reproit` metadata stub that replaces a secret value. Byte-identical
/// to `metadata()` in the Node SDK: type plus length, never the value.
private func reproitRedactedStub(_ value: Any) -> [String: Any] {
  var kind = "null"
  var length: Any = NSNull()
  switch value {
  case let text as String:
    kind = "string"
    length = text.count
  case let number as NSNumber:
    // Bool bridges to NSNumber in Foundation JSON, so check it first.
    if CFGetTypeID(number) == CFBooleanGetTypeID() {
      kind = "boolean"
    } else {
      kind = CFNumberIsFloatType(number) ? "number" : "integer"
    }
  case let items as [Any]:
    kind = "array"
    length = items.count
  case is [String: Any]:
    kind = "object"
  default:
    kind = "null"
  }
  return ["$reproit": ["redacted": true, "type": kind, "length": length]]
}

/// Recursive structural redaction matching the backend SDKs: secret-named
/// fields become a `$reproit` stub, everything else recurses. Shapes and
/// non-secret values survive, so a redacted capsule still replays.
func reproitRedactExchangeValue(_ value: Any) -> Any {
  if let items = value as? [Any] { return items.map(reproitRedactExchangeValue) }
  if let map = value as? [String: Any] {
    var result: [String: Any] = [:]
    for (key, child) in map {
      result[key] = reproitIsSecretField(key) ? reproitRedactedStub(child) : reproitRedactExchangeValue(child)
    }
    return result
  }
  return value
}

/// Bound one exchange body into the wire fields. A JSON body is parsed so the
/// structural redaction above sees fields rather than text; a body past the
/// inline budget keeps only its provable identity.
func reproitBoundedExchangeBody(_ data: Data?, contentType: String?) -> [String: Any] {
  guard let data, !data.isEmpty else { return [:] }
  if data.count > reproitMaxExchangeBodyBytes {
    var bounded: [String: Any] = ["bodyBytes": data.count, "truncated": true]
    if let digest = reproitSha256Hex(data) { bounded["bodySha256"] = digest }
    return bounded
  }
  if (contentType ?? "").lowercased().contains("application/json"),
    let parsed = try? JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])
  {
    return ["body": reproitRedactExchangeValue(parsed)]
  }
  guard let text = String(data: data, encoding: .utf8) else {
    // Non-UTF8 payload: identity only, never a lossy transcription.
    var bounded: [String: Any] = ["bodyBytes": data.count, "truncated": true]
    if let digest = reproitSha256Hex(data) { bounded["bodySha256"] = digest }
    return bounded
  }
  return ["body": text]
}

/// Lowercase, cap, and redact header values for one exchange side.
func reproitBoundedExchangeHeaders(_ headers: [String: String]) -> [String: Any] {
  // Sort before capping so the retained subset is deterministic run to run.
  let capped = headers.sorted { $0.key.lowercased() < $1.key.lowercased() }
    .prefix(reproitMaxExchangeHeaders)
  var result: [String: String] = [:]
  for (name, value) in capped {
    let key = name.lowercased()
    result[key] = reproitIsSecretField(key) ? "<reproit:secret>" : value
  }
  return result.isEmpty ? [:] : ["headers": result]
}

/// One captured production exchange, ready to nest in a capture batch.
struct ReproItExchange {
  let method: String
  let url: String
  let requestHeaders: [String: String]
  let requestBody: Data?
  let requestContentType: String?
  let status: Int
  let responseHeaders: [String: String]
  let responseBody: Data?
  let responseContentType: String?

  /// The wire object, field-for-field identical to the backend SDKs.
  func jsonObject() -> [String: Any] {
    var request: [String: Any] = ["method": method, "url": url]
    request.merge(reproitBoundedExchangeHeaders(requestHeaders)) { current, _ in current }
    request.merge(reproitBoundedExchangeBody(requestBody, contentType: requestContentType)) {
      current, _ in current
    }
    var response: [String: Any] = ["status": status]
    response.merge(reproitBoundedExchangeHeaders(responseHeaders)) { current, _ in current }
    response.merge(reproitBoundedExchangeBody(responseBody, contentType: responseContentType)) {
      current, _ in current
    }
    return ["protocol": "http", "request": request, "response": response]
  }
}

/// Process-wide store of production-captured exchanges.
///
/// Bounded and fail-open for the host app: recording never throws into the
/// caller, and overflow drops the OLDEST exchange rather than growing without
/// limit. The store is inert until ``enable()`` is called, which the SDK does
/// only when the app opts in through `ReproItConfig.captureExchanges`.
final class ReproItExchangeStore {
  static let shared = ReproItExchangeStore()

  /// Exchanges retained for the next failure occurrence. A failure ships the
  /// dependency calls that led to it, so this is a flight recorder, not a
  /// full session log.
  private static let maxRetained = 32

  private let lock = NSLock()
  private var enabled = false
  private var retained: [ReproItExchange] = []
  private(set) var droppedExchanges = 0

  /// Turn production recording on. Idempotent.
  func enable() {
    lock.lock()
    enabled = true
    lock.unlock()
  }

  var isEnabled: Bool {
    lock.lock()
    defer { lock.unlock() }
    return enabled
  }

  func record(_ exchange: ReproItExchange) {
    lock.lock()
    defer { lock.unlock() }
    guard enabled else { return }
    retained.append(exchange)
    if retained.count > Self.maxRetained {
      retained.removeFirst(retained.count - Self.maxRetained)
      droppedExchanges += 1
    }
  }

  /// The retained exchanges, oldest first.
  func snapshot() -> [ReproItExchange] {
    lock.lock()
    defer { lock.unlock() }
    return retained
  }

  /// Test and teardown hook.
  func reset() {
    lock.lock()
    enabled = false
    retained.removeAll()
    droppedExchanges = 0
    lock.unlock()
  }
}

File: sdk/reproit-android/src/main/kotlin/com/reproit/android/Exchange.kt
package com.reproit.android

import java.security.MessageDigest
import java.util.Locale

/**
 * Outbound dependency exchange capture, the Android port of
 * `sdk/reproit-backend-node/instrument.js`.
 *
 * An exchange is the request the app sent and the response the dependency
 * returned. It is what hermetic local replay serves, so responses are captured
 * verbatim up to a fixed inline budget; an over-budget body keeps only its byte
 * count and sha256 and is marked truncated, and replay fails closed on it with a
 * named reason instead of guessing.
 *
 * Pure Kotlin with no `android.*` import so every bound and every redaction rule
 * is host-testable; [CausalHttp] supplies the transport.
 *
 * Two redaction vocabularies live in this file on purpose:
 *   - [redactCausalValue] keeps the `<reproit:...>` placeholders the RUNNER path
 *     has always emitted, so runner output stays byte-identical.
 *   - [redactCaptureValue] emits Node's `$reproit` metadata stub, because the
 *     replay matcher treats a `$reproit` marker as "any value stood here" and
 *     would compare a `<reproit:...>` string literally and diverge.
 */

/** Inline body budget per exchange side, byte-identical to the Node SDK. */
internal const val MAX_EXCHANGE_BODY_BYTES = 8 * 1024

/** Recorded headers per exchange side, byte-identical to the Node SDK. */
internal const val MAX_EXCHANGE_HEADERS = 32

/** Captured exchanges retained for one failure occurrence; drops oldest. */
internal const val MAX_CAPTURED_EXCHANGES = 32

/**
 * Node's secret-key vocabulary (`SECRET_PARTS` in reproit-backend-node/index.js).
 * Deliberately separate from [causalSecretField]: this list adds
 * `idempotencykey`, and changing the runner list would change runner output.
 */
private val CAPTURE_SECRET_PARTS =
  listOf(
    "password",
    "passwd",
    "secret",
    "token",
    "authorization",
    "cookie",
    "email",
    "phone",
    "apikey",
    "publishablekey",
    "privatekey",
    "accesskey",
    "signingkey",
    "idempotencykey",
  )

/** Fold a field name the way Node does: drop non-alphanumerics, lowercase. */
internal fun captureSecretField(key: String): Boolean {
  val compact = key.filter { it.isLetterOrDigit() }.lowercase(Locale.ROOT)
  return CAPTURE_SECRET_PARTS.any(compact::contains)
}

/**
 * Recursive structural redaction with Node's `$reproit` placeholder shape.
 * Secret-named fields keep their type and length so replay can still match
 * positionally; every other value passes through.
 */
internal fun redactCaptureValue(value: Any?): Any? =
  when (value) {
    is List<*> -> value.map(::redactCaptureValue)
    is Map<*, *> ->
      value.entries.associate { (key, child) ->
        key.toString() to
          if (captureSecretField(key.toString())) captureMetadata(child)
          else redactCaptureValue(child)
      }
    else -> value
  }

/** Node's `metadata()`: the typed stub that replaces a secret-named value. */
private fun captureMetadata(value: Any?): Map<String, Any?> {
  val type =
    when (value) {
      null -> "null"
      is Boolean -> "boolean"
      is Int,
      is Long -> "integer"
      is Double -> if (value == value.toLong().toDouble()) "integer" else "number"
      is String -> "string"
      is List<*> -> "array"
      is Map<*, *> -> "object"
      else -> "null"
    }
  val length: Any? =
    when (value) {
      is String -> value.codePointCount(0, value.length)
      is List<*> -> value.size
      else -> null
    }
  return mapOf("\$reproit" to linkedMapOf("redacted" to true, "type" to type, "length" to length))
}

internal fun sha256Hex(bytes: ByteArray): String {
  val digest = MessageDigest.getInstance("SHA-256").digest(bytes)
  val out = StringBuilder(64)
  for (byte in digest) out.append("%02x".format(byte))
  return out.toString()
}

/**
 * Bound one exchange body. Returns an empty map for an absent or empty body, the
 * provable identity only when the body exceeds [MAX_EXCHANGE_BODY_BYTES], and
 * otherwise the parsed JSON (when the content type declares it) or the text.
 */
internal fun boundedBody(body: ByteArray?, contentType: String?): Map<String, Any?> {
  if (body == null || body.isEmpty()) return emptyMap()
  if (body.size > MAX_EXCHANGE_BODY_BYTES) {
    return linkedMapOf(
      "bodyBytes" to body.size,
      "bodySha256" to sha256Hex(body),
      "truncated" to true,
    )
  }
  val text = body.toString(Charsets.UTF_8)
  if (contentType != null && contentType.contains("application/json")) {
    try {
      // Mark decoded nulls as DATA so the encoder keeps them: a response of
      // {"prices": null} must replay as {"prices": null}, not as a body with
      // the key missing, which would reproduce a different error.
      return linkedMapOf("body" to markJsonNulls(Json.decode(text)))
    } catch (_: Throwable) {
      // Declared JSON that does not parse is recorded as text below.
    }
  }
  return linkedMapOf("body" to text)
}

/**
 * Recursively replace decoded JSON nulls with [JsonNull] so [Json] encodes
 * them instead of dropping the key. Applies only to captured payloads, never
 * to the SDK's own optional event fields.
 */
internal fun markJsonNulls(value: Any?): Any? =
  when (value) {
    null -> JsonNull
    is Map<*, *> -> {
      val marked = LinkedHashMap<String, Any?>(value.size)
      for ((k, v) in value) marked[k.toString()] = markJsonNulls(v)
      marked
    }
    is List<*> -> value.map { markJsonNulls(it) }
    else -> value
  }

/** Bound and lowercase one exchange side's headers. */
internal fun boundedHeaders(headers: Map<String, String>): Map<String, Any?> {
  if (headers.isEmpty()) return emptyMap()
  val bounded = LinkedHashMap<String, Any?>()
  // The cap is defined over NAME SORTED order, never arrival order. Go recorded
  // a different subset on every run because it capped a randomized map before
  // sorting it, and a capsule whose recorded headers vary between runs cannot be
  // matched twice. A LinkedHashMap does not randomize, so the wrong subset here
  // would be stable and therefore even harder to notice.
  val sorted =
    headers.entries
      .map { it.key.lowercase(Locale.ROOT) to it.value }
      .sortedBy { it.first }
      .take(MAX_EXCHANGE_HEADERS)
  for ((name, value) in sorted) bounded[name] = value
  return linkedMapOf("headers" to bounded)
}

/** Read one header case-insensitively, the way Node's `headerValue` does. */
internal fun headerValue(headers: Map<String, String>, name: String): String? =
  headers.entries.firstOrNull { it.key.equals(name, ignoreCase = true) }?.value

/**
 * Build one redacted `{protocol, request, response}` exchange in the exact shape
 * the Node and Rust SDKs emit, so one replay engine serves all three.
 */
internal fun captureHttpExchange(
  method: String,
  url: String,
  requestHeaders: Map<String, String>,
  requestBody: ByteArray?,
  status: Int,
  responseHeaders: Map<String, String>,
  responseBody: ByteArray?,
): Map<String, Any?> {
  val request = LinkedHashMap<String, Any?>()
  request["method"] = method.uppercase(Locale.ROOT)
  request["url"] = url
  request.putAll(boundedHeaders(requestHeaders))
  request.putAll(boundedBody(requestBody, headerValue(requestHeaders, "content-type")))
  val response = LinkedHashMap<String, Any?>()
  response["status"] = status
  response.putAll(boundedHeaders(responseHeaders))
  response.putAll(boundedBody(responseBody, headerValue(responseHeaders, "content-type")))
  val exchange =
    linkedMapOf<String, Any?>("protocol" to "http", "request" to request, "response" to response)
  @Suppress("UNCHECKED_CAST")
  return redactCaptureValue(exchange) as Map<String, Any?>
}

/**
 * The runner-path secret vocabulary. Unchanged from its original home in
 * [CausalHttp] so the runner wire stays byte-identical; moved here only so the
 * pure logic compiles and tests on the host JVM without the Android SDK.
 */
internal fun causalSecretField(key: String): Boolean {
  val compact =
    key.lowercase(Locale.ROOT).filterNot { it == '-' || it == '_' || it == '.' || it == ' ' }
  return listOf(
      "password",
      "passwd",
      "secret",
      "token",
      "authorization",
      "cookie",
      "email",
      "phone",
      "apikey",
      "publishablekey",
      "privatekey",
      "accesskey",
      "signingkey",
    )
    .any(compact::contains)
}

internal fun redactCausalValue(value: Any?): Any? =
  when (value) {
    is List<*> -> value.map(::redactCausalValue)
    is Map<*, *> ->
      value.entries.associate { (key, child) ->
        key.toString() to
          if (causalSecretField(key.toString())) causalTypedValue(child)
          else redactCausalValue(child)
      }
    else -> value
  }

private fun causalTypedValue(value: Any?): String =
  when (value) {
    null -> "<reproit:null>"
    is String -> "<reproit:string:length=${value.codePointCount(0, value.length)}>"
    is Boolean -> "<reproit:boolean>"
    is Number -> "<reproit:number>"
    is List<*> -> "<reproit:array:length=${value.size}>"
    is Map<*, *> -> "<reproit:object:keys=${value.size}>"
    else -> "<reproit:unknown>"
  }
